### PR TITLE
Add support for Yubikey 4 & NEO to cvmfs-server

### PR DESCRIPTION
**Pull request by DrDaveD**
See [CVM-1259](https://sft.its.cern.ch/jira/browse/CVM-1259)

### CODE REVIEW

File: cvmfs/server/cvmfs_server_masterkeycard.sh [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+# Check if a masterkeycard is available to be used
+# If not, the reason is sent to stdout and false is returned,
+# otherwise there's nothing to stdout and true is returned
+masterkeycard_available() {
+  local pattern="Yubikey.*CCID"
+  local reason=""
+  if ! lsusb | grep -q "$pattern"; then
+    reason="USB device matching \"$pattern\" not present"
+  elif ! which opensc-tool > /dev/null 2>&1; then
+    reason="opensc-tool (from opensc rpm) not found"
```
Comment (jblomer) [inline cvmfs/server/cvmfs_server_masterkeycard.sh]: This should also indicate the Debian package name

File: cvmfs/server/cvmfs_server_masterkeycard.sh [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+  elif ! which systemctl > /dev/null 2>&1; then
+    reason="masterkeycard only supported on systems with systemctl"
+  elif ! systemctl is-active --quiet pcscd.socket; then
+    reason="systemctl unit pcscd.socket is not active"
+  elif ! systemctl is-enabled --quiet pcscd.socket; then
+    reason="systemctl unit pcscd.socket is not enabled"
+  elif ! opensc-tool -l | grep -q "$pattern"; then
+    reason="opensc-tool -l has no device matching \"$pattern\""
+  elif ! which yubico-piv-tool >/dev/null 2>&1; then
+    reason="yubico-piv-tool (from yubico-piv-tool rpm) not found"
```
Comment (jblomer) [inline cvmfs/server/cvmfs_server_masterkeycard.sh]: Should also indicate the Debian package name

File: cvmfs/server/cvmfs_server_masterkeycard.sh [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+  elif yubico-piv-tool -a status 2>&1 | grep -q "Failed to connect"; then
+    reason="yubico-piv-tool failed to connect to device"
+  elif ! openssl engine pkcs11 2>/dev/null|grep -q "pkcs11 engine"; then
+    reason="openssl engine pkcs11 (from engine_pkcs11 rpm) not found"
```
Comment (jblomer) [inline cvmfs/server/cvmfs_server_masterkeycard.sh]: Should also indicate the Debian package name

File: cvmfs/server/cvmfs_server_masterkeycard.sh [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+          return 1
+        fi
+      fi
+
+      pubkey="`masterkeycard_read_pubkey`" || die "Failure reading pub key from mastercard"
+      for name in $goodnames; do
+        masterkey="/etc/cvmfs/keys/${name}.masterkey"
+        if cvmfs_sys_file_is_regular $masterkey; then
+          echo "Removing $masterkey and updating pub key"
+          rm -f $masterkey
```
Comment (jblomer) [inline cvmfs/server/cvmfs_server_masterkeycard.sh]: Should we use `shred` here instead of `rm`?

File: packaging/rpm/cvmfs-universal.spec [hunk 5]
```diff
@@ -163,6 +163,7 @@ Requires: httpd
 Requires: libcap
 Requires: lsof
 Requires: rsync
+Requires: usbutils
```
Comment (jblomer) [inline packaging/rpm/cvmfs-universal.spec]: We should add this dependency for the Debian and the Arch (aps) package, too.

**Comment by DrDaveD** [comment]
Ok I made your suggested changes.  I haven't fully tested on cvmfs_server but I know there's a problem with the openssl use of the pkcs11 engine because of this [Debian bug](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=815004).

**Comment by DrDaveD** [comment]
I plan to update the documentation but I prefer to wait until my cvmfs_server resign additions have also been merged.   I wasn't sure about when to update pre-release documentation.  Is it OK to update the master branch in doc-cvmfs at any time?

**Pull request merged**